worker/memory.go

```diff
@@ -0,0 +1,112 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.sia.tech/renterd/api"
+)
+
+type (
+	// memoryManager helps regulate processes that use a lot of memory. Such as
+	// uploads and downloads.
+	memoryManager struct {
+		totalAvailable uint64
+
+		mu        sync.Mutex
+		sigNewMem sync.Cond
+		available uint64
+	}
+
+	acquiredMemory struct {
+		mm *memoryManager
+
+		mu        sync.Mutex
+		remaining uint64
+	}
+)
+
+func newMemoryManager(maxMemory uint64) (*memoryManager, error) {
+	if maxMemory == 0 {
+		return nil, fmt.Errorf("maxMemory cannot be 0")
+	}
+	mm := &memoryManager{
+		totalAvailable: maxMemory,
+	}
+	mm.available = mm.totalAvailable
+	mm.sigNewMem = *sync.NewCond(&mm.mu)
+	return mm, nil
+}
+
+func (mm *memoryManager) Status() api.MemoryStatus {
+	mm.mu.Lock()
+	defer mm.mu.Unlock()
+	return api.MemoryStatus{
+		Available: mm.available,
+		Total:     mm.totalAvailable,
+	}
+}
+
+func (mm *memoryManager) AcquireMemory(ctx context.Context, amt uint64) <-chan *acquiredMemory {
+	if amt == 0 {
+		panic("cannot acquire 0 memory")
+	}
+	memChan := make(chan *acquiredMemory, 1)
+	// block until enough memory is available
```

Comment:
I think it's rather unfortunate we're not returning early when the context's got cancelled or its deadline was reached. Couldn't we do something like this:

```go
    memAvailChan := make(chan struct{})

    go func() {
        sigNewMem.L.Lock()
        defer sigNewMem.L.Unlock()
        for mm.avail < amt { 
            sigNewMem.Wait()
        }
        close(condChan)
    }()

    select {
    case <-ctx.Done():
        // yay we don't have to needlessly block the caller until we would have memory
        return ctx.Err()
    case <-memAvailChan:
        // yay we have memory
    }
```

Reply:
I didn't realise you were returning a channel, aren't you missing the goroutine? Seems like you wanted to insta-return a channel and forgot to block in a goroutine here